network-libp2p: Add fixes and improvements to discovery protocol

jsdanielh opened the pull request:
## What's in this pull request?

- Add verification of peer contacts received from the different messages that can transmit a list of peer contacts in the discovery protocol. The verification includes checks for appropriate lengths, signatures and number of addresses.
- Add timeout for the state transition in the discovery handler such that a peer don't hold us in a state indefinitely if it doesn't send us back a message we expect.
  This fixes #2458.
- Do not use the observed addresses sent through the discovery protocol since these addresses are not authenticated and it's been observed that they are not adding any useful address to a peer and instead were adding addresses with incorrect TCP/UDP ports.
- Reduce the reported observed addresses to a single address within the discovery protocol. This avoids DoS attacks from a peer sending several observed addresses for a specific connection. In terms of functionality, it stays the same since we were only reporting a single observed address (the address that got us a connection). Also reduced the interaction between the handler and the behaviour by setting handler attributes on the constructor instead of using events to the handler.
- Add limits to the peer contact addresses and add the ability to verify that these limits are followed when a peer contact is received from the network (discovery protocol).

## Pull request checklist

- [X] All tests pass. The project builds and runs.
- [X] I have resolved any merge conflicts.
- [X] I have resolved all `clippy` and `rustfmt` warnings.

jsdanielh committed:
network-libp2p: Add limits to the peer contact addresses

Add limits to the peer contact addresses and add the ability to
verify that these limits are followed when a peer contact is received
from the network (discovery protocol).

jsdanielh committed:
network-libp2p: Reduce the reported observed addresses to 1

Reduce the reported observed addresses to a single address within the
discovery protocol. This avoids DoS attacks from a peer sending
several observed addresses for a specific connection. In terms of
functionality, it stays the same since we were only reporting a
single observed address (the address that got us a connection).
Also reduced the interaction between the handler and the behaviour
by setting handler attributes on the constructor instead of using
events to the handler.

jsdanielh committed:
network-libp2p: Do not use the observed addresses

Do not use the observed addresses sent through the discovery protocol
since these addresses are not authenticated and it's been observed
that they are not adding any useful address to a peer and instead
were adding addresses with incorrect TCP/UDP ports.

jsdanielh committed:
network-libp2p: Add discovery state transition timeout

Add timeout for the state transition in the discovery handler such
that a peer don't hold us in a state indefinitely if it doesn't send
us back a message we expect.

This fixes #2458.

Eligioo commented on network-libp2p/src/discovery/peer_contacts.rs:
```diff
@@ -107,6 +125,15 @@ impl PeerContact {
             .retain(|addr| !to_remove_addresses.contains(addr));
     }
+
+    /// Verifies wether the lengths of the advertised addresses are within
```
```suggestion
    /// Verifies whether the lengths of the advertised addresses are within
```

nibhar commented on network-libp2p/src/discovery/handler.rs:
I cannot  comment on it directly, but in the match arm for `HandlerState::ReceiveHandshakeAck` it begins with a `peer_contact` and `peer_contacts` as well. The `peer_contact` is `verify`ed, but for the `peer_contacts` I cannot find it. They could have too many addresses as well as I believe they are also received over the network and thus are not correctly sized by construction. I very well may be missing something here though.

jsdanielh replied:
Good point. Added verification of the size and signatures in another commit.

jsdanielh committed:
network-libp2p: Verify also peer contacts received

Add verification of peer contacts received from the different
messages that can transmit a list of peer contacts in the discovery
protocol. The verification includes checks for appropriate lengths,
signatures and number of addresses.

hrxi commented on network-libp2p/src/discovery/handler.rs:
```diff
@@ -169,21 +165,23 @@ pub struct Handler {
 }
 
 impl Handler {
+    const STATE_TRANSITION_TIMEOUT_MS: u64 = 750;
```
This is quite a short timeout, some packet loss might already reach that. Not sure what would be a good value.

```suggestion
    const STATE_TRANSITION_TIMEOUT: Duration = Duration::from_millis(5000);
```

hrxi commented on network-libp2p/src/discovery/handler.rs:
```diff
@@ -374,6 +376,10 @@ impl ConnectionHandler for Handler {
                     }
 
                     self.state = HandlerState::ReceiveHandshake;
+                    self.state_timeout
+                        .as_mut()
+                        .expect("state timeout must exist")
+                        .reset(Duration::from_millis(Self::STATE_TRANSITION_TIMEOUT_MS));
```
Just another option without an `expect`. If you want to make sure there was a timeout for the previous state, too, then your version is better.
```suggestion
                    self.state_timeout = Some(Delay::new(Duration::from_millis(
                        Self::STATE_TRANSITION_TIMEOUT_MS,
                    )));
```

hrxi commented on network-libp2p/src/discovery/peer_contacts.rs:
```diff
@@ -97,8 +111,12 @@ impl PeerContact {
+            return Err(PeerContactError::AdvertisedAddressesExceeded);
+        }
+        self.addresses.extend(addresses);
+        Ok(())
```
Maybe this could add addresses until it overflows, so that at least some are recorded?

hrxi commented on network-libp2p/tests/network.rs:
```diff
@@ -352,9 +352,9 @@ async fn create_network_with_n_peers(n_peers: usize) -> Vec<Network> {
+#[test(tokio::test(flavor = "multi_thread", worker_threads = 10))]
 async fn connections_stress_and_reconnect() {
-    let peers: usize = 15;
+    let peers: usize = 10;
```
Is there a reason for this change?

jsdanielh replied:
No reason. Changed it back.

hrxi reviewed (commented):
Looks good to me, left some comments.

nibhar commented on network-libp2p/src/discovery/behaviour.rs:
```diff
@@ -318,11 +304,9 @@ impl NetworkBehaviour for Behaviour {
+                    .push_back(ToSwarm::NewExternalAddrCandidate(observed_address));
             }
             HandlerOutEvent::Update => self.events.push_back(ToSwarm::GenerateEvent(Event::Update)),
             HandlerOutEvent::Error(_) => self.events.push_back(ToSwarm::CloseConnection {
```
If I understand correctly this is where the `StateTransitionTimeout` would end up. I may have missed it, but I think some logging would be nice here.

nibhar reviewed (approved):
LGTM
